# BIG DATA PROCESSING

In [1]:
import findspark
findspark.init()

import pandas as pd
import pyspark

In [2]:
#Creamos una sesion de spark
from pyspark.sql import SparkSession

#Creamos un objeto de spark session
spark = SparkSession.builder.appName("pysparkdf").getOrCreate()

In [3]:
#Lectura del archivo world-happiness-report
data = spark.read.csv(
    "C:/Users/028cmf/Desktop/Nueva carpeta/KeepCoding/big-data-processing/PROYECTO FINAL/Datasets/world-happiness-report.csv",
    sep = ',',
    header = True,)

data.printSchema()

root
 |-- Country name: string (nullable = true)
 |-- year: string (nullable = true)
 |-- Life Ladder: string (nullable = true)
 |-- Log GDP per capita: string (nullable = true)
 |-- Social support: string (nullable = true)
 |-- Healthy life expectancy at birth: string (nullable = true)
 |-- Freedom to make life choices: string (nullable = true)
 |-- Generosity: string (nullable = true)
 |-- Perceptions of corruption: string (nullable = true)
 |-- Positive affect: string (nullable = true)
 |-- Negative affect: string (nullable = true)



In [4]:
#Cambiar la estructura de los datos
from pyspark.sql.types import *

data_schema = [
               StructField('Country name', StringType(), True),
               StructField('year', IntegerType(), True),
               StructField('Life Ladder', DoubleType(), True),
               StructField('Log GDP per capita', DoubleType(), True),
               StructField('Social support', DoubleType(), True),
               StructField('Healthy life expectancy at birth', DoubleType(), True),
               StructField('Freedom to make life choices', DoubleType(), True),
               StructField('Generosity', DoubleType(), True),
               StructField('Perceptions of corruption', DoubleType(), True),
               StructField('Positive affect', DoubleType(), True),
               StructField('Negative affect', DoubleType(), True),
            ]

final_struc = StructType(fields=data_schema)

In [5]:
data = spark.read.csv(
    "C:/Users/028cmf/Desktop/Nueva carpeta/KeepCoding/big-data-processing/PROYECTO FINAL/Datasets/world-happiness-report.csv",
    sep = ',',
    header = True,
    schema = final_struc)

data.printSchema()

root
 |-- Country name: string (nullable = true)
 |-- year: integer (nullable = true)
 |-- Life Ladder: double (nullable = true)
 |-- Log GDP per capita: double (nullable = true)
 |-- Social support: double (nullable = true)
 |-- Healthy life expectancy at birth: double (nullable = true)
 |-- Freedom to make life choices: double (nullable = true)
 |-- Generosity: double (nullable = true)
 |-- Perceptions of corruption: double (nullable = true)
 |-- Positive affect: double (nullable = true)
 |-- Negative affect: double (nullable = true)



In [6]:
data.show(5)

+------------+----+-----------+------------------+--------------+--------------------------------+----------------------------+----------+-------------------------+---------------+---------------+
|Country name|year|Life Ladder|Log GDP per capita|Social support|Healthy life expectancy at birth|Freedom to make life choices|Generosity|Perceptions of corruption|Positive affect|Negative affect|
+------------+----+-----------+------------------+--------------+--------------------------------+----------------------------+----------+-------------------------+---------------+---------------+
| Afghanistan|2008|      3.724|              7.37|         0.451|                            50.8|                       0.718|     0.168|                    0.882|          0.518|          0.258|
| Afghanistan|2009|      4.402|              7.54|         0.552|                            51.2|                       0.679|      0.19|                     0.85|          0.584|          0.237|
| Afghanistan|2

In [7]:
#Lectura del archivo world-happiness-report-2021
data_2021 = spark.read.csv(
    "C:/Users/028cmf/Desktop/Nueva carpeta/KeepCoding/big-data-processing/PROYECTO FINAL/Datasets/world-happiness-report-2021.csv",
    sep = ',',
    header = True,)

data_2021.printSchema()

root
 |-- Country name: string (nullable = true)
 |-- Regional indicator: string (nullable = true)
 |-- Ladder score: string (nullable = true)
 |-- Standard error of ladder score: string (nullable = true)
 |-- upperwhisker: string (nullable = true)
 |-- lowerwhisker: string (nullable = true)
 |-- Logged GDP per capita: string (nullable = true)
 |-- Social support: string (nullable = true)
 |-- Healthy life expectancy: string (nullable = true)
 |-- Freedom to make life choices: string (nullable = true)
 |-- Generosity: string (nullable = true)
 |-- Perceptions of corruption: string (nullable = true)
 |-- Ladder score in Dystopia: string (nullable = true)
 |-- Explained by: Log GDP per capita: string (nullable = true)
 |-- Explained by: Social support: string (nullable = true)
 |-- Explained by: Healthy life expectancy: string (nullable = true)
 |-- Explained by: Freedom to make life choices: string (nullable = true)
 |-- Explained by: Generosity: string (nullable = true)
 |-- Explained 

In [8]:
#Cambiar la estructura de los datos
from pyspark.sql.types import *

data_schema_2021 = [
               StructField('Country name', StringType(), True),
               StructField('Regional indicator', StringType(), True),
               StructField('Ladder score', DoubleType(), True),
               StructField('Standard error of lader score', DoubleType(), True),
               StructField('upperwhisker', DoubleType(), True),
               StructField('lowerwhisker', DoubleType(), True),
               StructField('Logged GDP per capita', DoubleType(), True),
               StructField('Social support', DoubleType(), True),
               StructField('Healthy life expectancy', DoubleType(), True),
               StructField('Freedom to make life choices', DoubleType(), True),
               StructField('Generosity', DoubleType(), True),
               StructField('Perceptions of corruption', DoubleType(), True),
               StructField('Ladder score in Dystopia', DoubleType(), True),
               StructField('Explained by: log GDP per capita', DoubleType(), True),
               StructField('Explained by: Social support', DoubleType(), True),
               StructField('Explained by: Healthy life expectancy', DoubleType(), True),
               StructField('Explained by: Freedom to make life choices', DoubleType(), True),
               StructField('Explained by: Generosity', DoubleType(), True),
               StructField('Explained by: Perceptions of corruption', DoubleType(), True),
               StructField('Dystopia + residual', DoubleType(), True),
            ]

final_struc_2021 = StructType(fields=data_schema_2021)

In [9]:
data_2021 = spark.read.csv(
    "C:/Users/028cmf/Desktop/Nueva carpeta/KeepCoding/big-data-processing/PROYECTO FINAL/Datasets/world-happiness-report-2021.csv",
    sep = ',',
    header = True,
    schema = final_struc_2021)

data_2021.printSchema()

root
 |-- Country name: string (nullable = true)
 |-- Regional indicator: string (nullable = true)
 |-- Ladder score: double (nullable = true)
 |-- Standard error of lader score: double (nullable = true)
 |-- upperwhisker: double (nullable = true)
 |-- lowerwhisker: double (nullable = true)
 |-- Logged GDP per capita: double (nullable = true)
 |-- Social support: double (nullable = true)
 |-- Healthy life expectancy: double (nullable = true)
 |-- Freedom to make life choices: double (nullable = true)
 |-- Generosity: double (nullable = true)
 |-- Perceptions of corruption: double (nullable = true)
 |-- Ladder score in Dystopia: double (nullable = true)
 |-- Explained by: log GDP per capita: double (nullable = true)
 |-- Explained by: Social support: double (nullable = true)
 |-- Explained by: Healthy life expectancy: double (nullable = true)
 |-- Explained by: Freedom to make life choices: double (nullable = true)
 |-- Explained by: Generosity: double (nullable = true)
 |-- Explained b

In [10]:
data_2021.show(5)

+------------+------------------+------------+-----------------------------+------------+------------+---------------------+--------------+-----------------------+----------------------------+----------+-------------------------+------------------------+--------------------------------+----------------------------+-------------------------------------+------------------------------------------+------------------------+---------------------------------------+-------------------+
|Country name|Regional indicator|Ladder score|Standard error of lader score|upperwhisker|lowerwhisker|Logged GDP per capita|Social support|Healthy life expectancy|Freedom to make life choices|Generosity|Perceptions of corruption|Ladder score in Dystopia|Explained by: log GDP per capita|Explained by: Social support|Explained by: Healthy life expectancy|Explained by: Freedom to make life choices|Explained by: Generosity|Explained by: Perceptions of corruption|Dystopia + residual|
+------------+------------------+-

## Ejercicio 1

¿Cuál es el país más “feliz” del 2021 según la data? (considerar que la columna “Ladder score” mayor número más feliz es el país)

In [11]:
data_2021.select(["Country name", "Regional indicator","Ladder score"]).orderBy("Ladder score", ascending = False).show(1)

+------------+------------------+------------+
|Country name|Regional indicator|Ladder score|
+------------+------------------+------------+
|     Finland|    Western Europe|       7.842|
+------------+------------------+------------+
only showing top 1 row



Por tanto, el país más feliz del 2021 fue Finlandia.

## Ejercicio 2

¿Cuál es el país más “feliz” del 2021 por continente según la data?

In [12]:
from pyspark.sql.window import Window
from pyspark.sql.functions import row_number,desc
windowSpec  = Window.partitionBy("Regional indicator").orderBy(desc("Ladder score"))

data_2021.select(["Country name", "Regional indicator","Ladder score"]).withColumn("row_number",row_number().over(windowSpec)) \
    .filter("row_number = 1") \
    .show(truncate = False)

+------------------------+----------------------------------+------------+----------+
|Country name            |Regional indicator                |Ladder score|row_number|
+------------------------+----------------------------------+------------+----------+
|Czech Republic          |Central and Eastern Europe        |6.965       |1         |
|Uzbekistan              |Commonwealth of Independent States|6.179       |1         |
|Taiwan Province of China|East Asia                         |6.584       |1         |
|Costa Rica              |Latin America and Caribbean       |7.069       |1         |
|Israel                  |Middle East and North Africa      |7.157       |1         |
|New Zealand             |North America and ANZ             |7.277       |1         |
|Nepal                   |South Asia                        |5.269       |1         |
|Singapore               |Southeast Asia                    |6.377       |1         |
|Mauritius               |Sub-Saharan Africa          

## Ejercicio 3

¿Cuál es el país que más veces ocupó el primer lugar en todos los años?

In [13]:
from pyspark.sql.window import Window
from pyspark.sql.functions import row_number,desc
windowSpec  = Window.partitionBy("year").orderBy(desc("Life Ladder"))

data.select(["Country name", "year","Life Ladder"]).withColumn("row_number",row_number().over(windowSpec)) \
    .filter("row_number = 1") \
    .show(truncate = False)

+------------+----+-----------+----------+
|Country name|year|Life Ladder|row_number|
+------------+----+-----------+----------+
|Denmark     |2005|8.019      |1         |
|Finland     |2006|7.672      |1         |
|Denmark     |2007|7.834      |1         |
|Denmark     |2008|7.971      |1         |
|Denmark     |2009|7.683      |1         |
|Denmark     |2010|7.771      |1         |
|Denmark     |2011|7.788      |1         |
|Switzerland |2012|7.776      |1         |
|Canada      |2013|7.594      |1         |
|Denmark     |2014|7.508      |1         |
|Norway      |2015|7.603      |1         |
|Finland     |2016|7.66       |1         |
|Finland     |2017|7.788      |1         |
|Finland     |2018|7.858      |1         |
|Finland     |2019|7.78       |1         |
|Finland     |2020|7.889      |1         |
+------------+----+-----------+----------+



In [14]:
df = data.select(["Country name", "year","Life Ladder"]).withColumn("row_number",row_number().over(windowSpec)) \
    .filter("row_number = 1")
df.groupBy("Country name").count().show()

+------------+-----+
|Country name|count|
+------------+-----+
|     Finland|    6|
|      Norway|    1|
|     Denmark|    7|
| Switzerland|    1|
|      Canada|    1|
+------------+-----+



Por tanto, el país que más veces ocupó el primer puesto de felicidad fue Dinamarca.

## Ejercicio 4

¿Qué puesto de Felicidad tiene el país con mayor GDP del 2020?

In [15]:
from pyspark.sql.functions import filter
data_2020 = data.select(["Country name","year", "Life Ladder", "Log GDP per capita"]).filter(data.year == 2020)
data_2020.show()

+--------------------+----+-----------+------------------+
|        Country name|year|Life Ladder|Log GDP per capita|
+--------------------+----+-----------+------------------+
|             Albania|2020|      5.365|             9.497|
|           Argentina|2020|      5.901|              9.85|
|           Australia|2020|      7.137|             10.76|
|             Austria|2020|      7.213|            10.851|
|             Bahrain|2020|      6.173|             10.62|
|          Bangladesh|2020|       5.28|             8.472|
|             Belgium|2020|      6.839|            10.771|
|               Benin|2020|      4.408|             8.102|
|             Bolivia|2020|      5.559|             8.998|
|Bosnia and Herzeg...|2020|      5.516|             9.583|
|              Brazil|2020|       6.11|             9.522|
|            Bulgaria|2020|      5.598|             9.991|
|            Cambodia|2020|      4.377|             8.362|
|            Cameroon|2020|      5.241|             8.17

In [16]:
from pyspark.sql.window import Window
from pyspark.sql.functions import row_number, desc

windowSpecLifeLadder = Window.partitionBy("year").orderBy(desc("Life Ladder"))
windowSpecGDP = Window.partitionBy("year").orderBy(desc("Log GDP per capita"))

data_2020.withColumn("row_number_LifeLadder", row_number().over(windowSpecLifeLadder)) \
    .withColumn("row_number_GDP", row_number().over(windowSpecGDP)) \
    .show(truncate=False)

+--------------------+----+-----------+------------------+---------------------+--------------+
|Country name        |year|Life Ladder|Log GDP per capita|row_number_LifeLadder|row_number_GDP|
+--------------------+----+-----------+------------------+---------------------+--------------+
|Ireland             |2020|7.035      |11.323            |13                   |1             |
|Switzerland         |2020|7.508      |11.081            |4                    |2             |
|United Arab Emirates|2020|6.458      |11.053            |27                   |3             |
|Norway              |2020|7.29       |11.042            |8                    |4             |
|United States       |2020|7.028      |11.001            |14                   |5             |
|Denmark             |2020|7.515      |10.91             |3                    |6             |
|Netherlands         |2020|7.504      |10.901            |5                    |7             |
|Austria             |2020|7.213      |1

En 2020, el país con mayor GDP per capita fue Irlanda y ocupó el puesto número 13 de felicidad.

## Ejercicio 5

¿En que porcentaje ha variado a nivel mundial el GDP promedio del 2020 respecto al 2021? ¿Aumentó o disminuyó?

In [32]:
from pyspark.sql.functions import avg, round

average_gdp_2020 = data_2020.agg(round(avg("Log GDP per capita"), 2).alias("Average GDP 2020"))
average_gdp_2020.show()

+----------------+
|Average GDP 2020|
+----------------+
|            9.75|
+----------------+



In [28]:
from pyspark.sql.functions import avg, round

average_gdp_2021 = data_2021.select("Logged GDP per capita").agg(round(avg("Logged GDP per capita"),2).alias("Average GDP 2021"))
average_gdp_2021.show()

+----------------+
|Average GDP 2021|
+----------------+
|            9.43|
+----------------+



In [41]:
percentage_decrease = ((average_gdp_2020.select("Average GDP 2020").collect()[0][0] - average_gdp_2021.select("Average GDP 2021").collect()[0][0]) / average_gdp_2020.select("Average GDP 2020").collect()[0][0]) * 100
print("Percentage Decrease:", percentage_decrease)

Percentage Decrease: 3.2820512820512846


Como podemos observar, del 2020 al 2021 ha disminuido un 3.28% pasando de 9.75 a 9.43.

## Ejercicio 6

¿Cuál es el país con mayor expectativa de vida (“Healthy life expectancy at birth”)? Y ¿Cuánto tenia en ese indicador en el 2019?

In [18]:
data_2021.describe("Healthy life expectancy").show()

+-------+-----------------------+
|summary|Healthy life expectancy|
+-------+-----------------------+
|  count|                    149|
|   mean|      64.99279865771811|
| stddev|      6.762043090404319|
|    min|                 48.478|
|    max|                 76.953|
+-------+-----------------------+



In [19]:
data.describe("Healthy life expectancy at birth").show()

+-------+--------------------------------+
|summary|Healthy life expectancy at birth|
+-------+--------------------------------+
|  count|                            1894|
|   mean|              63.359373812037965|
| stddev|               7.510244618236356|
|    min|                            32.3|
|    max|                            77.1|
+-------+--------------------------------+



In [20]:
data.select(["Country name", "year", "Healthy life expectancy at birth"]).orderBy("Healthy life expectancy at birth", ascending = False).show(1)

+------------+----+--------------------------------+
|Country name|year|Healthy life expectancy at birth|
+------------+----+--------------------------------+
|   Singapore|2019|                            77.1|
+------------+----+--------------------------------+
only showing top 1 row



Comparando esta variable en ambos dataframes, observamos que el máximo es 77.1 y corresponde a Singapore en el año 2019.